zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/compensation/CompensationEventExecutionTest.java

```diff
@@ -2329,6 +2331,501 @@ public void shouldDeleteSubscriptionInTerminatedSubprocessMultiInstanceScope() {
+                                    compensation.serviceTask("Undo-B").zeebeJobType("Undo-B")))
+                .serviceTask(
+                    "C",
+                    task ->
+                        task.zeebeJobType("C")
+                            .boundaryEvent()
+                            .compensation(
+                                compensation ->
+                                    compensation.serviceTask("Undo-C").zeebeJobType("Undo-C")))
+                .serviceTask("D", task -> task.zeebeJobType("D"));
+
+    final BpmnModelInstance process =
+        Bpmn.createExecutableProcess(PROCESS_ID)
+            .startEvent()
+            .serviceTask(
+                "A",
+                task ->
+                    task.zeebeJobType("A")
+                        .boundaryEvent()
+                        .compensation(
+                            compensation ->
+                                compensation.serviceTask("Undo-A").zeebeJobType("Undo-A")))
+            .subProcess("subprocess", subprocessBuilder)
+            .endEvent()
+            .done();
+
+    final EndEvent endEvent = process.getModelElementById("compensation-throw-event");
+    endEvent.builder().compensateEventDefinition().activityRef("B");
+
+    ENGINE.deployment().withXmlResource(process).deploy();
+
+    final long processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
+
+    ENGINE.job().ofInstance(processInstanceKey).withType("A").complete();
+    ENGINE.job().ofInstance(processInstanceKey).withType("B").complete();
+    ENGINE.job().ofInstance(processInstanceKey).withType("C").complete();
+
+    // when
+    ENGINE.job().ofInstance(processInstanceKey).withType("D").withErrorCode("error").throwError();
+
+    // when
+    ENGINE.job().ofInstance(processInstanceKey).withType("Undo-B").complete();
+
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted())
+        .extracting(r -> r.getValue().getElementId(), Record::getIntent)
+        .containsSubsequence(
+            tuple("event-subprocess", ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple("compensation-throw-event", ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple("Undo-B", ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple("compensation-throw-event", ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple("event-subprocess", ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple("subprocess", ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(PROCESS_ID, ProcessInstanceIntent.ELEMENT_COMPLETED))
+        .doesNotContain(
+            tuple("Undo-A", ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple("Undo-C", ProcessInstanceIntent.ELEMENT_ACTIVATED));
+  }
+
+  @Test
+  public void shouldTriggerCompensationForActivityInsideEventSubprocess() {
```

Comment:
💭 We should add another test case to verify that, if a referenced activity is not completed, the compensation is not triggered